File: chrome_bridge.js
// Native Chrome automation. Always address a process, window and tab by ID.
function run(argv) {
    const request = JSON.parse(argv[0]);
    let stage = request.action;
    function fail(code, message) { return {ok: false, code: code, error: message}; }
    function tabInfo(t) { return {id: Number(t.id()), url: t.url() || ""}; }
    try {
        const app = Application(request.pid);
        if (!app.running()) return JSON.stringify(fail("BROWSER_CLOSED", "Chrome process has exited."));
        const windows = app.windows();
        if (request.action === "inventory") {
            return JSON.stringify({ok: true, value: windows.map(w => ({
                id: Number(w.id()), tabs: w.tabs().map(tabInfo)
            }))});
        }
        if (request.action === "new_window") {
            const before = new Set(windows.map(w => Number(w.id())));
            // Chrome can create the window but return an invalid object
            // specifier to JXA. Recover by observing IDs, never by creating twice.
            let createError = "";
            try { app.windows.push(app.Window()); }
            catch (e) { createError = String(e); }
            stage = "new_window.locate";
            const deadline = Date.now() + 3000;
            while (Date.now() < deadline) {
                const created = app.windows().find(w => !before.has(Number(w.id())));
                if (created && created.tabs().length) {
                    const wid = Number(created.id());
                    const w = app.windows.byId(String(wid));
                    const t = w.tabs.byId(String(w.tabs()[0].id()));
                    stage = "new_window.navigate";
                    t.url = request.url;
                    return JSON.stringify({ok: true, value: {id: wid, tab: tabInfo(t)}});
                }
                delay(0.05);
            }
            return JSON.stringify(fail("BROWSER_CLOSED", "Chrome did not create a tracking tab. " + createError));
        }
        if (!windows.some(w => Number(w.id()) === request.window_id))
            return JSON.stringify(fail("BROWSER_CLOSED", "The tracking Chrome window was closed."));
        const w = app.windows.byId(String(request.window_id));
        let value = null;
        if (request.action === "tabs") value = w.tabs().map(tabInfo);
        else if (request.action === "close_window") { w.close(); value = true; }
        else if (request.action === "new_tab") {
            const before = new Set(w.tabs().map(t => Number(t.id())));
            try { w.tabs.push(app.Tab({url: request.url})); } catch (e) { /* Check the created ID below. */ }
            const deadline = Date.now() + 3000;
            while (Date.now() < deadline) {
                const t = w.tabs().find(t => !before.has(Number(t.id())));
                if (t) { value = tabInfo(t); break; }
                delay(0.05);
            }
            if (!value) return JSON.stringify(fail("TAB_NOT_FOUND", "Chrome did not create a result tab."));
        } else {
            const tabs = w.tabs();
            const index = tabs.findIndex(t => Number(t.id()) === request.tab_id);
            if (index < 0) return JSON.stringify(fail("TAB_NOT_FOUND", "The tracking tab was closed or is no longer available."));
            const t = w.tabs.byId(String(request.tab_id));
            if (request.action === "evaluate") value = t.execute({javascript: request.script});
            else if (request.action === "tab") value = tabInfo(t);
            else if (request.action === "navigate") { t.url = request.url; value = tabInfo(t); }
            else if (request.action === "close_tab") { t.close(); value = true; }
            else if (request.action === "activate" || request.action === "bounds") {
                w.activeTabIndex = index + 1;
                w.index = 1;
                app.activate();
                value = request.action === "bounds" ? w.bounds() : true;
            } else return JSON.stringify(fail("NAVIGATION", "Unknown Chrome bridge action."));
        }
        return JSON.stringify({ok: true, value: value === undefined ? null : value});
    } catch (e) {
        // Localized JXA errors can collapse an Apple Events denial to only
        // "Error: 发生错误。".  Preserve the numeric field so -1743 is
        // classified correctly on every macOS language.
        const errorNumber = Number(e && e.errorNumber);
        const suffix = Number.isFinite(errorNumber) ? " (" + errorNumber + ")" : "";
        const message = stage + ": " + String(e) + suffix;
        // Only an explicit Chrome/OS permission error is a permission failure.
        const denied = errorNumber === -1743 || errorNumber === -10004
            || /javascript.*apple\s*(events?|事件|script)|apple\s*(events?|事件|script).*javascript|not authorized|not permitted|未获授权|不允许访问|(-1743)|(-10004)/i.test(message);
        return JSON.stringify(fail(denied ? "BROWSER_PERMISSION" : "NAVIGATION", message));
    }
}

File: chrome_bridge.m
#import <AppKit/AppKit.h>
#import <Foundation/Foundation.h>
#import <ScriptingBridge/ScriptingBridge.h>

@protocol ChromeGenericMethods
- (void)close;
- (id)executeJavascript:(NSString *)javascript;
@end

@interface ChromeApplication : SBApplication
- (SBElementArray *)windows;
@end

@interface ChromeWindow : SBObject <ChromeGenericMethods>
- (SBElementArray *)tabs;
- (NSString *)id;
@property NSInteger index;
@property NSRect bounds;
@property NSInteger activeTabIndex;
@property(copy, readonly) id activeTab;
@end

@interface ChromeTab : SBObject <ChromeGenericMethods>
- (NSString *)id;
@property(copy) NSString *URL;
@end

@interface BridgeDelegate : NSObject <SBApplicationDelegate>
@property(strong) NSError *lastError;
@end

@implementation BridgeDelegate
- (id)eventDidFail:(const AppleEvent *)event withError:(NSError *)error {
    self.lastError = error;
    return nil;
}
@end

static NSString *clean(NSString *value) {
    NSString *text = value ?: @"";
    text = [text stringByReplacingOccurrencesOfString:@"\t" withString:@"\\t"];
    text = [text stringByReplacingOccurrencesOfString:@"\r" withString:@"\\r"];
    return [text stringByReplacingOccurrencesOfString:@"\n" withString:@"\\n"];
}

static void ok(NSString *body) {
    printf("OK\n%s\n", [(body ?: @"") UTF8String]);
}

static void fail(NSString *code, NSString *stage, NSString *message, NSInteger number) {
    NSString *body = [NSString stringWithFormat:@"ERROR\t%@\t%@: %@ (%ld)\n",
                      code, stage, clean(message), (long)number];
    fputs([body UTF8String], stdout);
}

static BOOL reportDelegateError(BridgeDelegate *delegate, NSString *stage) {
    NSError *error = delegate.lastError;
    if (!error) return NO;
    NSString *code = @"NAVIGATION";
    if (error.code == -1743 || error.code == -10004) code = @"BROWSER_PERMISSION";
    else if ([stage isEqualToString:@"window"]) code = @"BROWSER_CLOSED";
    else if ([stage isEqualToString:@"tab"]) code = @"TAB_NOT_FOUND";
    fail(code, stage, error.localizedDescription, error.code);
    return YES;
}

static ChromeWindow *findWindow(ChromeApplication *app, long long target) {
    for (ChromeWindow *window in [app windows]) {
        if ([[window id] longLongValue] == target) return window;
    }
    return nil;
}

static ChromeTab *findTab(ChromeWindow *window, long long target, NSInteger *index) {
    NSInteger current = 0;
    for (ChromeTab *tab in [window tabs]) {
        current += 1;
        if ([[tab id] longLongValue] == target) {
            if (index) *index = current;
            return tab;
        }
    }
    return nil;
}

int main(int argc, const char *argv[]) {
    @autoreleasepool {
        if (argc < 6) {
            fail(@"NAVIGATION", @"arguments", @"Missing Chrome bridge arguments.", -1);
            return 0;
        }
        NSString *action = [NSString stringWithUTF8String:argv[1]];
        pid_t pid = (pid_t)strtol(argv[2], NULL, 10);
        long long windowId = strtoll(argv[3], NULL, 10);
        long long tabId = strtoll(argv[4], NULL, 10);
        NSString *payload = [NSString stringWithUTF8String:argv[5]];
        NSString *stage = action;
        @try {
            ChromeApplication *app = (ChromeApplication *)[SBApplication applicationWithProcessIdentifier:pid];
            BridgeDelegate *delegate = [BridgeDelegate new];
            app.delegate = delegate;
            app.timeout = 15 * 60;
            if (!app || !app.running) {
                fail(@"BROWSER_CLOSED", stage, @"The selected Chrome process is not running.", -600);
                return 0;
            }

            if ([action isEqualToString:@"inventory"]) {
                NSMutableArray<NSString *> *lines = [NSMutableArray array];
                for (ChromeWindow *window in [app windows]) {
                    for (ChromeTab *tab in [window tabs]) {
                        [lines addObject:[NSString stringWithFormat:@"%@\t%@\t%@",
                                          [window id], [tab id], clean(tab.URL)]];
                    }
                }
                if (reportDelegateError(delegate, stage)) return 0;
                ok([lines componentsJoinedByString:@"\n"]);
                return 0;
            }

            if ([action isEqualToString:@"new_window"]) {
                Class windowClass = [app classForScriptingClass:@"window"];
                ChromeWindow *window = [[windowClass alloc] initWithProperties:@{}];
                [[app windows] addObject:window];
                window = [window get];
                ChromeTab *tab = window.activeTab;
                tab.URL = payload;
                if (reportDelegateError(delegate, stage)) return 0;
                ok([NSString stringWithFormat:@"%@\t%@\t%@", [window id], [tab id], clean(tab.URL)]);
                return 0;
            }

            stage = @"window";
            ChromeWindow *window = findWindow(app, windowId);
            if (!window) {
                fail(@"BROWSER_CLOSED", stage, @"The tracking Chrome window was closed.", -1728);
                return 0;
            }
            if ([action isEqualToString:@"tabs"]) {
                NSMutableArray<NSString *> *lines = [NSMutableArray array];
                for (ChromeTab *tab in [window tabs]) {
                    [lines addObject:[NSString stringWithFormat:@"%@\t%@", [tab id], clean(tab.URL)]];
                }
                if (reportDelegateError(delegate, stage)) return 0;
                ok([lines componentsJoinedByString:@"\n"]);
                return 0;
            }
            if ([action isEqualToString:@"close_window"]) {
                [window close];
                if (reportDelegateError(delegate, stage)) return 0;
                ok(@"true");
                return 0;
            }
            if ([action isEqualToString:@"new_tab"]) {
                Class tabClass = [app classForScriptingClass:@"tab"];
                ChromeTab *tab = [[tabClass alloc] initWithProperties:@{@"URL": payload}];
                [[window tabs] addObject:tab];
                tab = [tab get];
                if (reportDelegateError(delegate, @"new_tab")) return 0;
                ok([NSString stringWithFormat:@"%@\t%@", [tab id], clean(tab.URL)]);
                return 0;
            }

            stage = @"tab";
            NSInteger tabIndex = 0;
            ChromeTab *tab = findTab(window, tabId, &tabIndex);
            if (!tab) {
                fail(@"TAB_NOT_FOUND", stage, @"The tracking tab was closed or is no longer available.", -1728);
                return 0;
            }
            if ([action isEqualToString:@"evaluate"]) {
                stage = @"evaluate";
                id value = [tab executeJavascript:payload];
                if (reportDelegateError(delegate, stage)) return 0;
                ok(value ? [value description] : @"");
                return 0;
            }
            if ([action isEqualToString:@"tab"]) {
                ok([NSString stringWithFormat:@"%@\t%@", [tab id], clean(tab.URL)]);
                return 0;
            }
            if ([action isEqualToString:@"navigate"]) {
                tab.URL = payload;
                if (reportDelegateError(delegate, stage)) return 0;
                ok([NSString stringWithFormat:@"%@\t%@", [tab id], clean(tab.URL)]);
                return 0;
            }
            if ([action isEqualToString:@"close_tab"]) {
                [tab close];
                if (reportDelegateError(delegate, stage)) return 0;
                ok(@"true");
                return 0;
            }
            if ([action isEqualToString:@"activate"] || [action isEqualToString:@"bounds"]) {
                window.activeTabIndex = tabIndex;
                window.index = 1;
                [app activate];
                if (reportDelegateError(delegate, stage)) return 0;
                if ([action isEqualToString:@"bounds"]) {
                    NSRect rect = window.bounds;
                    ok([NSString stringWithFormat:@"%.0f,%.0f,%.0f,%.0f",
                        rect.origin.x, rect.origin.y,
                        rect.origin.x + rect.size.width, rect.origin.y + rect.size.height]);
                } else {
                    ok(@"true");
                }
                return 0;
            }
            fail(@"NAVIGATION", action, @"Unknown Chrome bridge action.", -1708);
        } @catch (NSException *exception) {
            NSString *code = [stage isEqualToString:@"window"] ? @"BROWSER_CLOSED" :
                             [stage isEqualToString:@"tab"] ? @"TAB_NOT_FOUND" : @"NAVIGATION";
            fail(code, stage, exception.reason, -1);
        }
    }
    return 0;
}
